server/src/main/java/org/elasticsearch/search/aggregations/AggregationErrors.java

```diff
@@ -21,4 +27,89 @@ private AggregationErrors() {}
     public static IllegalArgumentException invalidPathElement(String element, String path) {
         return new IllegalArgumentException("Invalid path element [" + element + "] in path [" + path + "]");
     }
+
+    /**
+     * This error indicates that an aggregation is being used on a value type that it is not capable of processing, such as taking
+     * the sum of a keyword. It is a 400 class error and should not be retried.
+     *
+     * This is a legacy version of this error; in general, we should aim to use the
+     * {@link org.elasticsearch.search.aggregations.support.ValuesSourceType} version below
+     *
+     * @param valuesSource The values source we resolved from the query
+     * @param name The name of the aggregation
+     * @return an appropriate exception type
+     */
+    public static IllegalArgumentException unsupportedValuesSourceType(ValuesSource valuesSource, String name) {
+        return new IllegalArgumentException(
+            "ValuesSource type [" + valuesSource.toString() + "] is not supported for aggregation [" + name + "]"
+        );
+    }
+
+    /**
+     * This error indicates that a rate aggregation is being invoked without a single Date Histogram parent, as is required.  This is a
+     * 400 class error and should not be retried.
+     *
+     * @param name the name of the rate aggregation
+     * @return an appropriate exception
+     */
+    public static RuntimeException rateWithoutDateHistogram(String name) {
+        return new IllegalArgumentException(
+            "aggregation ["
+                + name
+                + "] does not have exactly one date_histogram value source; exactly one is required when using with rate aggregation"
+        );
+    }
+
+    /**
+     * This error indicates that the backing indices for a field have different, incompatible, types (e.g. IP and Keyword).  This
+     * causes a failure at reduce time, and is not retryable (and thus should be a 400 class error)
+     *
+     * @param aggregationName - The name of the aggregation
+     * @param position - optional, for multisource aggregations.  Indicates the position of the field causing the problem.
+     * @return - an appropriate exception
+     */
+    public static RuntimeException reduceTypeMissmatch(String aggregationName, Optional<Integer> position) {
```

Comment:
nit: typo - s/Missmatch/Mismatch/

Reply:
English is my native language, and I still never know when it wants to double a letter or not.  I didn't see this until after it merged, I'll fix in the next iteration.